### DISCUSSION TITLE
Keeping #batches consistent across multiple graphs in link prediction when minibatching

### DISCUSSION THREAD

**Discussion by jasperhyp** [pyg-team/pytorch_geometric#3745] [Q&A] [answered]
Hi there,  I am doing a link prediction task on multiple graphs.  Those graphs are of the same kind, but have distinct numbers of nodes/edges ranging from ~1000 edges to ~30000 edges (graphs are fed to the model in the form of a dictionary).  I do not need to do minibatching now, but the graphs may expand later to ~100,000 edges, so it's better to start minibatching now.  To properly train the model, it might be better if I can include samples from all graphs in each minibatch.  My plan is that I use `GraphSAINTEdgeSampler` to partition each graph into a fixed number of minibatches, and then construct a list of dictionary, which I can then iterate through.  

However, I found that `GraphSAINTEdgeSampler` seems to sample edges repetitively (say for one graph I have ~10000 edges, it is sampling ~20000 in total when I sample some random batch size, and ~5000 for some smaller batch size...), and there is no explicit way to fix the number of batches (with flexible batch size across graphs).  Thanks in advance for your suggestions!

P.S. I didn't plan to use `DataLoader` & `RandomLinkSplit` as is the usual case for link prediction and batching because it does not seem like DataLoader can properly handle my situation of multiple graphs.

**Comment by rusty1s** [comment, marked as answer]
* `GraphSAINTEdgeSampler` should never sample edges twice. Do you have a small script to reproduce this?

* You are right that `GraphSAINTEdgeSampler` does not fix the number of batches. This is due to how GraphSAINT is defined, i.e. nodes are potentially re-sampled across iterations.

* I'm not entirely sure why you need `GraphSAINTEdgeSampler` in the first place. Processing 100k edges should be easily doable in full-batch mode, and you can even use `DataLoader` to stack multiple graphs together if GPU memory allows.

**Reply by jasperhyp** [reply]
Hi Matthias--   thanks for your reply!  And sorry for my late reply.  I checked the results and the "sampling twice" issue is mainly something with my negative sample generator function, which repetitively samples some node pairs as negative none-links.  Still, I don't completely understand how to partition the total sample set with it:

```{python}
> (x, x.shape)
(tensor([[ 0.2552, -0.4089,  3.4416,  ...,  0.3084,  1.1509,  1.0713],
         [ 0.1826,  0.1775,  1.6552,  ...,  1.1285,  0.4492, -0.9804],
         [ 0.7314, -0.0076,  1.2123,  ..., -0.2218, -1.7908,  0.0506],
         ...,
         [ 0.8964,  0.1391,  2.6470,  ..., -0.0838, -0.6572, -0.5200],
         [ 0.2860,  0.4352,  0.6064,  ..., -1.0620,  1.4289,  0.2552],
         [ 1.8244,  1.4329,  1.3471,  ...,  0.9960, -1.4072, -2.2303]],
        device='cuda:0'),
 torch.Size([2373, 2048]))
> (data_train, data_train['total_edge_index'].shape)
({'total_edge_index': tensor([[1031, 1031, 1031,  ..., 1756, 2255,  101],
          [2226, 1107, 1956,  ...,  782,  457,  130]], device='cuda:0'),
  'total_edge_type': tensor([3, 3, 3,  ..., 3, 3, 3], device='cuda:0'),
  'y': tensor([1., 1., 1.,  ..., 0., 0., 0.], device='cuda:0')},
 torch.Size([2, 28902]))
> total_data = Data(x=x, edge_index=data_train['total_edge_index'].to("cpu"), 
     edge_attr=data_train['total_edge_type'], y=data_train['y'])

> loader = GraphSAINTEdgeSampler(train_data, batch_size=500, num_steps=10, sample_coverage=100)
Compute GraphSAINT normalization: : 237300it [01:56, 2045.76it/s] 

> for batch in loader:
       print(batch)
Data(num_nodes=519, edge_index=[2, 9783], x=[519, 2048], edge_attr=[9783], y=[9783], node_norm=[519], edge_norm=[9783])
Data(num_nodes=488, edge_index=[2, 9390], x=[488, 2048], edge_attr=[9390], y=[9390], node_norm=[488], edge_norm=[9390])
Data(num_nodes=512, edge_index=[2, 9767], x=[512, 2048], edge_attr=[9767], y=[9767], node_norm=[512], edge_norm=[9767])
Data(num_nodes=496, edge_index=[2, 9878], x=[496, 2048], edge_attr=[9878], y=[9878], node_norm=[496], edge_norm=[9878])
Data(num_nodes=526, edge_index=[2, 10016], x=[526, 2048], edge_attr=[10016], y=[10016], node_norm=[526], edge_norm=[10016])
```

If I am not misunderstanding, it seems to give me `num_steps` batches (~subgraphs?) with ~`batch_size` nodes.  Obviously some edges are sampled twice, as the sum of numbers of edges across `Data` objects exceeds the original size (28902).

You are right that processing 100k edges should be easily doable in full-batch mode.  Actually I'm not entirely sure how large the dataset might scale to, maybe 50k nodes and 500k edges for one graph, but the problem is that all those graphs need to be put together in each batch (because I am combining the information from those graphs in the model per iteration), and that number of graphs could easily go beyond 100-ish.  So I am trying to sample a few links from each graph respectively and put them together every batch, and keep the total number of batches across all sampling schemes on those graphs the same, so that in each batch links from all graphs are present. 

In this case, using `DataLoader` might not yield what I want (maybe it can work if I transform all downstream dictionary operations to `scatter`-based operations, but many lines of codes need to be rewritten...), and I'll need probably a subgraph sampler that can fix the number of batches.  I have two scenarios, one is I need to sample subgraphs rather than random links, in this case edge-based samplers like GraphSAINTSampler are needed (and then simply use `zip(*list_of_loaders)` to pass 100-ish subgraphs into one batch training), but I am just not sure if the `num_steps` `Data` objects it generates can cover all edges in a graph in each epoch.  The second is that I just randomly sample links (well... better not do this).

**Reply by jasperhyp** [reply]
After a brief search I found the only sampler that can ensure all edges are sampled in at least one batch and allows number of batches to be easily controlled is -- NeighborSampler

**Reply by rusty1s** [reply]
This is not necessiarly true. `NeighborSampler`/`NeighborLoader` will start sampling for each node in the graph exactly once during each epoch, but due to the sampling it is not guaranteed that every edge is sampled once. I think what is currently missing in PyG is some form of `EdgeSampler` for tasks such as link prediction. Currently, the only option to achieve this is to inherit from `NeighborLoader` and modify the `input_nodes` to ensure that we start sampling for every `(src_node, dst_node)` combination (rather than for every node).

**Reply by rusty1s** [reply]
We can probably create an issue for that and lay out the details. Let me know if you have any interesting in working with us on this one :)

**Reply by jasperhyp** [reply]
> This is not necessiarly true. `NeighborSampler`/`NeighborLoader` will start sampling for each node in the graph exactly once during each epoch, but due to the sampling it is not guaranteed that every edge is sampled once. I think what is currently missing in PyG is some form of `EdgeSampler` for tasks such as link prediction. Currently, the only option to achieve this is to inherit from `NeighborLoader` and modify the `input_nodes` to ensure that we start sampling for every `(src_node, dst_node)` combination (rather than for every node).

Thanks for your follow-up, Matthias!  I thought I achieved the goal of sampling every node once because I set the params `num_neighbors=-1` and `replace=False` for `NeighborLoader`.  In this case, from what I understood in the document, I guess I am sampling every node once?  

```python
> train_data_0 = Data(x=x, edge_index=data_train.edge_index, edge_attr=data_train.edge_type, y=data_train.y)
> train_data_0.n_id = torch.arange(train_data_0.num_nodes)
> train_data_0
  Data(x=[2373, 2048], edge_index=[2, 28902], edge_attr=[28902], y=[28902], n_id=[2373])
> loader = NeighborLoader(train_data_0, num_neighbors=[-1], batch_size=int(np.ceil(train_data_0.num_nodes/10)), input_nodes=torch.arange(train_data_0.num_nodes))
> ls = []
> for batch in loader:
...   ori_map = batch.n_id.numpy()
...   mapped_edge_index = torch.tensor([[ori_map[i] for i in batch.edge_index.numpy()[0]], 
                                          [ori_map[i] for i in batch.edge_index.numpy()[1]]])
...   batch.total_edge_index = mapped_edge_index
...   ls.append(mapped_edge_index)
...   print(batch)
  Data(x=[1144, 2048], edge_index=[2, 2450], edge_attr=[2450], y=[2450], batch_size=238, total_edge_index=[2, 2450])
  Data(x=[1098, 2048], edge_index=[2, 2172], edge_attr=[2172], y=[2172], batch_size=238, total_edge_index=[2, 2172])
  Data(x=[1171, 2048], edge_index=[2, 2874], edge_attr=[2874], y=[2874], batch_size=238, total_edge_index=[2, 2874])
  Data(x=[1188, 2048], edge_index=[2, 3218], edge_attr=[3218], y=[3218], batch_size=238, total_edge_index=[2, 3218])
  Data(x=[1187, 2048], edge_index=[2, 3164], edge_attr=[3164], y=[3164], batch_size=238, total_edge_index=[2, 3164])
  Data(x=[1238, 2048], edge_index=[2, 3146], edge_attr=[3146], y=[3146], batch_size=238, total_edge_index=[2, 3146])
  Data(x=[1233, 2048], edge_index=[2, 3306], edge_attr=[3306], y=[3306], batch_size=238, total_edge_index=[2, 3306])
  Data(x=[1215, 2048], edge_index=[2, 3168], edge_attr=[3168], y=[3168], batch_size=238, total_edge_index=[2, 3168])
  Data(x=[1189, 2048], edge_index=[2, 3172], edge_attr=[3172], y=[3172], batch_size=238, total_edge_index=[2, 3172])
  Data(x=[1037, 2048], edge_index=[2, 2232], edge_attr=[2232], y=[2232], batch_size=231, total_edge_index=[2, 2232])
> len([(pair[0], pair[1]) for ts in ls for pair in ts.T.numpy()])  # number of all edges in batches, should be 28902
  28902
> len(set([(pair[0], pair[1]) for ts in ls for pair in ts.T.numpy()]))  # number of all unique edges in batches
  28170
> len(set([(pair[0], pair[1]) for pair in train_data_0.edge_index.T.numpy()]))  # number of all unique edges in original data
  28170
```

The number of edges in each batch sums to 28902.  As long as the "seed nodes" in every sampling batch are distinct, the edges are also distinct.  I didn't modify the negative sampler generator code, so the number of unique edges is smaller than the number of edges in the original training data.

PS.  I am very happy to contribute to a `NeighborLoader`-based `EdgeSampler`, if we can lay out the details clearly as you suggested.

**Reply by rusty1s** [reply]
Yes, your observation is correct, but it does indeed require `num_neighbors=-1` in the first iteration :)

**Reply by jasperhyp** [reply]
> Yes, your observation is correct, but it does indeed require `num_neighbors=-1` in the first iteration :)

Thanks Matthias!  `NeighborSampler` can somewhat be used as an edge sampler in this case ;)

Also, sorry to bother you on Christmas but I do have an additional question related to minibatching that I am hoping to seek advice.  We're using some GAT-based network to transform the embeddings of nodes from some 1024 dimension to 16 dimensional in full training mode (we can't modify this because of some downstream requirements), but after introducing minibatching, the already updated 16-dimensional node embeddings are fed into the 1024-in GAT again, which obviously causes an error.  We are thinking about either randomly initializing the node embeddings again or keep using the same original node embedding in each batch.  The problem is the goal is to learn the node embeddings and I really want it to get updated during each batch.  Do you happen to know some common practice for this issue?

**Reply by rusty1s** [reply]
Why do want to re-process already processed node embeddings once again? Input and output dimensionalities are fixed, and so you have to keep the original node embeddings for mini-batching as well.

**Reply by jasperhyp** [reply]
> Why do want to re-process already processed node embeddings once again? Input and output dimensionalities are fixed, and so you have to keep the original node embeddings for mini-batching as well.

We are learning the node embeddings, so they are updated (re-processed) every batch.  I'm also currently using the original node embeddings for each minibatch, and use the processed node embeddings after the final minibatch of the final epoch as the final node embedding to extract.  Hopefully the results won't be hugely affected compared with full-batch training.

**Reply by rusty1s** [reply]
So you are saying you apply the GNN+`NeighborSampler` strategy without any learning taking place? There is no need to do this IMO. `NeighborSampler` is well able to learn model parameters end-to-end.

**Reply by jasperhyp** [reply]
> So you are saying you apply the GNN+`NeighborSampler` strategy without any learning taking place? There is no need to do this IMO. `NeighborSampler` is well able to learn model parameters end-to-end.

Ahhh not really though.  We are learning GNN parameters throughout, and the embeddings transformed through the model from original input in each minibatch would be refined as training goes on.  Link prediction is used as a task to get some loss.  The model itself is end-to-end, and `NeighborSampler` is just one way to sample links for minibatching.  Looking back again, I guess it looks like a reasonable strategy?

**Reply by rusty1s** [reply]
Got it. As far as I understand, you are using link prediction as a self-supervision loss to learn node embeddings. This is indeed a reasonable strategy. Importantly, you need to keep original input node features for training your link prediction model.
